### PR TITLE
EC2 implementation (basic operations)

### PR DESCRIPTION
**Pull request by sashaafm**
Hello, 

Like I had said before I also want to contribute with the EC2 API (at least most used/needed calls, the API is huge). Like the RDS implementation I'm trying to mimic the other implementations as close as possible and trying to follow the guidelines. 

Still Work in Progress but I've already implemented some of the Instance related actions.

### CODE REVIEW

**Comment by sashaafm** [comment]
43 out of 200 actions/methods are now implemented.

**Comment by sashaafm** [comment]
I believe these are all the most useful/needed functions for EC2 usage.

**Comment by benwilson512** [comment]
This is a heroic effort indeed! Thank you so much for taking the time to handle this massive API. The same notes from the RDS PR generally apply here.

**Comment by sashaafm** [comment]
Yeah I had no idea it would be so big ahah After I correct the RDS PR and I'm sure it's correct I'll work on correcting this PR.

**Comment by sashaafm** [comment]
@benwilson512 everything done based on your notes :100:

File: lib/ex_aws/ec2/client.ex [hunk 1]
```diff
@@ -0,0 +1,1267 @@
...
+    network_interface_attachment_status                 :: (:attaching | :attached | :detaching | :detached)                           | 
+    network_interface_attachment_attach_time            :: binary                                                                      | 
+    network_interface_attachment_delete_on_termination  :: boolean                                                                     | 
+    network_interface_addresses_primary                 :: binary                                                                      | 
+    network_interface_addresses_association_public_ip   :: binary                                                                      | 
+    network_interface_addresses_association_ip_owner_id :: binary                                                                      | 
+    association_public_ip                               :: binary                                                                      | 
+    association_allocation_id                           :: binary                                                                      | 
+    association_association_id                          :: binary
```
Comment (benwilson512) [inline lib/ex_aws/ec2/client.ex]: This is a ridiculous amount of stuff, thank you so much!!!

**Comment by benwilson512** [comment]
I'm gonna have to look through this in more detail later, there is simply too much stuff here to go through at once. Looks good so far though.

**Comment by sashaafm** [comment]
Hi @benwilson512, when possible could you please review this and the RDS PR again? Thanks :+1:

File: lib/ex_aws/ec2/impl.ex [hunk 2]
```diff
@@ -0,0 +1,1009 @@
...
+
+  defp list_builder([h | []], key, count, state) do
+    Map.put_new(state, "#{key}.#{count}", h)
+  end
+
+  defp list_builder([h | t], key, count, state) do
+    list_builder t, key, count + 1, Map.put_new(state, "#{key}.#{count}", h)
+  end
+
+  defp list_builder_key_val([_h = {f, s} | []], key, count, state) do
```
Comment (benwilson512) [inline lib/ex_aws/ec2/impl.ex]: Why have `_h =` in here at all?

File: lib/ex_aws/ec2/impl.ex [hunk 3]
```diff
@@ -0,0 +1,1009 @@
...
+  end
+
+  defp normalize_opts(opts) do
+    opts
+    |> Enum.into(%{})
+    |> camelize_keys
+  end 
+
+  defp list_builder([h | []], key, count, state) do
+    Map.put_new(state, "#{key}.#{count}", h)
```
Comment (benwilson512) [inline lib/ex_aws/ec2/impl.ex]: put_new really isn't necessary here or in the subsequent areas. Anyone who managed to have this added to the state already is doing something wrong, and the data should be clobbered.

File: lib/ex_aws/ec2/impl.ex (hunk 2 above)
Comment (sashaafm) [inline lib/ex_aws/ec2/impl.ex]: I'm just used to doing it like that for expressiveness and explicit purposes (to make it clearer to someone new to the code that it's the head of the list). I'll remove it.

File: lib/ex_aws/ec2/impl.ex (hunk 3 above)
Comment (sashaafm) [inline lib/ex_aws/ec2/impl.ex]: Noted.

**Comment by benwilson512** [comment]
Have you had a chance to test any of these functions out live?

**Comment by sashaafm** [comment]
Yes. I've been using them in another project of mine [Nomad](https://github.com/sashaafm/nomad)

I've been using almost all functions from RDS and about 10-15 of the functions from EC2 and they've been working without problems. I've been using a branch that is a merge of the RDS and EC2 branches.

**Comment by benwilson512** [comment]
That is fantastic. I might tweak a thing or two but this looks largely good to go with me. Are you able to handle the merge issues or do you need help there?

**Comment by sashaafm** [comment]
I'll first correct a couple of things I noticed from my merged branch and then I'll try to merge and resolve the issues myself. If I run into difficulties I'll get back to you :) :+1: Thanks!

**Comment by benwilson512** [comment]
Closing in favor of https://github.com/CargoSense/ex_aws/pull/167

**Pull request closed without merging**